### PR TITLE
Fix item picker suggests to select items user does not have write access to

### LINKED ISSUE #15280: Users with READ collection permissions shouldn't be offered menu actions they can't execute
**Describe the bug**
Users with READ collection permissions can see menu action items that they don't have permissions to execute. Users are lead to believe that they CAN do a certain action which leads to all sorts of confusion (see: #13912).
Also, error messages are not uniform. Sometimes users will get inline error message, and sometimes they will be taken to a new screen with "Sorry, you don't have permissions to see that" message.

**Logs**
Usually just `403`.

**To Reproduce**
_As a user with read only permissions in a (root) collection go to that collection:_

#### Dashboards
1. Click on `...` menu for a dashboard 
2. You will see "Duplicate this item"
    - Modal will appear
    - When you try to "Duplicate", inline error is shown
image

##### Open a dashboard `/dashboard/:id` from a collection you have only read access to
1. Click on `...`
2. You will see:
    - Change title and description
        - Modal will appear
        - Change title or a description
        - Click "Update"
        - You'll be taken to a "Sorry, you don't have permissions to see that" page
    - Revision history
    - Duplicate
        - Modal will appear
        - When you try to "Duplicate", inline error is shown
    - Archive
        - Modal will appear
        - When you click "Archive", you'll be taken to a "Sorry, you don't have permissions to see that" page

#### Questions
- The `...` menu is not shown in a root collection. That's correct behavior.

##### Open a question `/question/:id` from a collection you have only read access to
1. Click on pencil icon next to the question's name
2. You will see:
    -  View revision history
    - Add to dashboard
        - Modal will appear where you can choose a dashboard to save this question to
        - If you choose a dashboard from a collection you have curate access to, there is no error and you can add that question
        - If you choose a dashboard from a collection you have only read access to
            - Question seems to be added
            - "You're editing this dashboard" will appear
            - You can resize the card and do everything just like you would normally when you add a question
            - But when you click "Save" button, it will turn into "Save failed"

**Expected behavior**
- We shouldn't display action items that user cannot execute
- We should hide collections that users can't save to
- We should unify error responses

**Information about your Metabase Installation:**
- Metabase version: v0.38.3-SNAPSHOT
- Metabase hosting environment: local dev, release-x.38.x
- Metabase internal database: H2 (default), sample dataset

**Severity**
P3 in terms of impact, but this happens in far too many places in the UI and that makes it more "serious".

### LINKED ISSUE #15281: Users with read permissions are offered to save new dashboard in collections they only have read access to
**Describe the bug**
Users with read permissions are offered to save new dashboard in collections they only have read access to. That leads to confusion and an error.
NOTE:
- The UI is initially correct and such users see only "My personal collection" but the error happens when they use a search functionality (see repro steps below)
- However, they can use the search and if type the name of a collection that lives inside a root (that they have read access to)
- The collection will be shown/offered

**Logs**
`403` error in the console.

**To Reproduce**
1. Click on `+` icon in the header and choose "New dashboard"
2. Modal will be shown - enter some name for the dashboard
3. You should see your personal collection 
4. Click on the selection menu - it will open a dropdown
5. Click on the "loupe" icon to be able to search
6. Search for any collection that lives inside the root collection (given that you have read access to it)
7. You will see that collection offered
8. Click on it - it will now be selected as a "go to collection"
9. Click "Create" button
10. Inline error will be shown

**Expected behavior**
Search shouldn't return collections users don't have curate access to

**Screenshots**
Initial UI
image

After searching for a collection user has read access to, and trying to save in it
image

**Information about your Metabase Installation:**
- Metabase version: v0.38.3-SNAPSHOT
- Metabase hosting environment: local dev, release-x.38.x
- Metabase internal database: H2 (default), sample dataset

**Severity**
P3

**Additional context**
Realated to #14052 and to #15280

### PR DESCRIPTION
**Pull request by kulyk**
Fixes #15281 

### Problem

There is a number of flows when a user adds-creates something somewhere and they need to specify the "target location". For example:

* select a dashboard to save question to
* select a collection to save dashboard to

Users might have different permissions to these resources. Right now, when we display e.g. a list of collections, we just show the whole list. So it's possible to select a collection you won't be able to actually modify and you'll end up with an error somewhere on UI.

This PR ensures we filter this things according to users' permissions

Also fixes a bug when selecting a collection mentioned [here](https://github.com/metabase/metabase/issues/15281#issuecomment-804112093)

### CODE REVIEW

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:12]: Awesome, thanks for this comment. It's super helpful for the future reference.

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:303]: Why do we skip every test for "nodata" user?

Comment (kulyk) [reply frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:303]: Ouch, haven't noticed it. Removing, thanks!

**Review by flamber — COMMENTED** [review]
As for #15280:

This seems to remove the `...` picker menu for collection view-user, so they cannot duplicate a dashboard (and question in the future, 15255).
Also removed the option, when going to a dashboard as view-user. For questions, this option hasn't been visible yet, but it should.

A view-user should be able to duplicate items from a view-collection and save it to a collection they have curate-access to.
A view-user should not see options that can change the item in a view-collection like "Archive" or "Change title and description" (or "Move" etc).

**Comment by flamber** [comment]
As for #15281:

Going to a collection that is revoked or view-only, and click "New dashboard", then it still offers to save to that location by default.
image

**Comment by flamber** [comment]
But it seems to fix #8996

File: frontend/src/metabase/containers/ItemPicker.jsx
Comment (paulrosenzweig) [inline frontend/src/metabase/containers/ItemPicker.jsx:79]: We don't use this naming convention with leading underscores much.

File: frontend/src/metabase/containers/ItemPicker.jsx
Comment (paulrosenzweig) [inline frontend/src/metabase/containers/ItemPicker.jsx:102]: What about collections that are themselves writable, have children, but none of the children are writable? It seems like the logic should be `this is writable || this has writable children`. I haven't tested it though so I might be missing something.

**Review by paulrosenzweig — APPROVED** [review]
Code looks good! I left one question about behavior, but I didn't spot anything besides that.

**Comment by paulrosenzweig** [comment]
> A view-user should be able to duplicate items from a view-collection and save it to a collection they have curate-access to.

I think that change was done [here](https://github.com/metabase/metabase/pull/15573/files#diff-39f914fbd0b225abdcaa6f8848dadeb49298c1efec4cc759f8f9ee91f2c91a2cR310).

**Comment by kulyk** [comment]
> Going to a collection that is revoked or view-only, and click "New dashboard", then it still offers to save to that location by default

@flamber thanks for catching that 💪 Will fix

**Comment by kulyk** [comment]
> > A view-user should be able to duplicate items from a view-collection and save it to a collection they have curate-access to.
> 
> I think that change was done [here](https://github.com/metabase/metabase/pull/15573/files#diff-39f914fbd0b225abdcaa6f8848dadeb49298c1efec4cc759f8f9ee91f2c91a2cR310).

Yes, you're right, I think it should be addressed in another PR then

File: frontend/src/metabase/containers/ItemPicker.jsx
Comment (kulyk) [reply frontend/src/metabase/containers/ItemPicker.jsx]: You're right, they get hidden in this case. Good catch, thank you!

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:50]:
```suggestion
                  cy.get(".AdminSelect").findByText("Second collection");
```
No need to use explicit `.should("exist")` assertion. `cy.findByText()` fails if it doesn't find a string and/or if it finds multiple strings so implicit assertion is always preferred.

Btw, I suggested we check for `.AdminSelect` because it is more specific and it makes it more clear to anyone who reads this test later what was the intention behind this assertion.

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:350]: All of this can and should be a single test without an additional describe block. Again, the same suggestion applies as in the comment before - no need to explicitly use`.should("exist")`.

If you take a look at the test runner sidebar and the generated titles, you'll see that we don't have to explicitly say "user has `write` access to". Every member from `curate` group has write access to all collections.
image


Also, I don't see any benefit in using/testing the search in this particular test because "Orders in a dashboard" was already shown in the modal. Testing search makes more sense for users who might not have write access to some collection because we want to make sure they are not even offered to them.

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (kulyk) [reply frontend/test/metabase/scenarios/collections/permissions.cy.spec.js]: Thanks for your feedback! Applied the change in my own commit not to trigger the CI once again

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (kulyk) [reply frontend/test/metabase/scenarios/collections/permissions.cy.spec.js]: Awesome! Merged the tests, removed explicit `"exists"` assertion, removed the "has write access" notes

**Comment by kulyk** [comment]
@nemanjaglumac thanks for your feedback, I guess everything is addressed and fixed now. Please take a look again

File: frontend/test/metabase/scenarios/question/native.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/native.cy.spec.js:232]: Everything looks good. Why was this change needed?

**Review by nemanjaglumac — APPROVED** [review]
Left a comment for `native.cy.spec.js` but overall looks pretty good. Thank you for addressing all the comments! 🙏

File: frontend/test/metabase/scenarios/question/native.cy.spec.js
Comment (kulyk) [reply frontend/test/metabase/scenarios/question/native.cy.spec.js:232]: That's great you asked, I was not 100% sure and went to double-check and there is actually an unexpected behavior here

**Comment by kulyk** [comment]
@paulrosenzweig @nemanjaglumac @flamber, thanks for the feedback you already shared. I found and fixed a few edge-cases here. Can you please give this PR one more look?

A bit of context on the recent changes:

After some fixes, [one of the native query tests failed](https://app.circleci.com/pipelines/github/metabase/metabase/15162/workflows/cafe15e3-6733-4479-a9a8-baa08e0c1942/jobs/580745/tests#failed-test-0). Here is the scenario of the failed test:

* a normal user creates a native query they will use later via the "Saved Questions" menu
* the user saves the query with a dialog changed in this PR
* the dialog has to suggest to save the question to the user's personal collection, but it suggests "Our analytics."



⬆️ This is a screenshot of the correct behavior. In the failing scenario, the popup suggests saving the question to "Our analytics".

The regression happened because of a change in [`getInitialCollectionId` selector](https://github.com/metabase/metabase/blob/e53c482a35dca1894413cbefa99f27cfafbee9fa/frontend/src/metabase/entities/collections.js#L91). When a user has to select a dashboard somewhere, the initial value is usually received from this selector. It collects a list of collection IDs in prioritized order (e.g., the top priority is the collection ID passed to the component via props, then via router params, etc. If nothing is passed, the selector fallbacks to the user's collection). Once the list is ready, we try to return the highest-priority option (closest to the top of the list).

This PR changes the selector, so we now check if a user has to write access to the collection to avoid irrelevant suggestions. That means, when the selector is called, **collections have to be present in the Redux store** to perform the check.

The test failed because it covers the behavior when `getInitialCollectionId` is called before the collection list is retrieved and saved to Redux. In this particular scenario, `getInitialCollectionId` would return user’s personal collection ID. But the selector will now take the collection object by ID to check its `can_write` flag. As the collections dictionary is empty, we can't check the `can_write,` and the selector will return `undefined`, making the modal fallback to the root collection ("Our analytics")

Ideally, we should fix the back-end API so the front end can retrieve only collections with write access. But for now, I wrapped collection selection input with `Collection.loadList` to guarantee each time user is offered to select a collection, and Redux store contains a list of collections. Also, I had to fix a few forms here and there, so they refresh when `initialCollectionId` is changed.

Please take a look and LMK if there is an easier way to achieve the same behavior 🙏

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:419]: Nice additional check!  🙏

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:482]: Nitpick: `. DashboardHeader` conveys more meaning. I had no idea whatsoever what is `.Entity` so I had to build the test and observe what's going on in the Cypress runner.

Don't have to change it in this PR. Just a general note.

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:448]: This is not "saving a question in a collection". We're adding a question to the dashboard that lives in a collection we have 'read' access to. The difference between the two scenarios is huge.

**Review by nemanjaglumac — APPROVED** [review]
Left a few comments, but tests overall look good.
Thank you for working hard on this! 

Didn't review the frontend, so best wait for other reviews as well.

**Review by paulrosenzweig — APPROVED** [review]
Left some comments, but it looks good!

One thought: Looking at this again, I'm somewhat nervous about unanticipated bugs. It touches a lot of important modals. Given that it's not a super urgent bug to fix, it might be worth shipping this with 40 so we have more time to catch anything.

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (kulyk) [reply frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:482]: Fixed, it's much better now. Thank you 🙏

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js
Comment (kulyk) [reply frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:448]: My bad. Found and fixed one more inconsistent test name. Thank you!

File: frontend/src/metabase/collections/components/CollectionCopyEntityModal.jsx
Comment (kulyk) [reply frontend/src/metabase/collections/components/CollectionCopyEntityModal.jsx:37]: Absolutely! Thanks, missed that

**Comment by kulyk** [comment]
@paulrosenzweig 

> Looking at this again, I'm somewhat nervous about unanticipated bugs. It touches a lot of important modals. Given that it's not a super urgent bug to fix, it might be worth shipping this with 40 so we have more time to catch anything.

Agree here. I'll rebase it on master them and share a heads up as the PR is merged

**Pull request merged**
